### DISCUSSION TITLE
Compose Enhancements

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#4004] [Development]
**To do**:

- [x] Add config item to toggle the visibility of custom headers
- [x] libcompose - allowing us to split up compose.c -- See 9b348bef8
      Create `compose/config.c` for the config
- [x] Add notification for `my_hdr` changes
- [x] Add observer for `my_hdr` changes
      This will allow the user to add/remove headers at will
- [ ] Move the function data to become Window data
- [x] Eliminate config global variables

**Comment by matthewhughes934** [comment]
Looks like fun, thanks for splitting it nicely into a list!

**Comment by flatcap** [comment]
I'm having some second thoughts about how my `my_hdr` command changes would work.

I liked the idea of composing an email, then adding a custom header at the press of a button.
Tying that into the `my_hdr` command gets a bit confusing though.

`my_hdr` and `unmy_hdr` have a permanence about them -- they affect all future emails.

Perhaps the `my_hdr` observers are still useful, but we also need a `<compose-add-one-off-header>` function.
We'd need a remove function too, which would need a menu to select _which_ header you mean.
Hmm...

**Comment by flatcap** [comment]
For the next couple, you'll need to **create** a new notification type.

If you look at the Event type, subtype and payload in [enum NotifyType](https://github.com/neomutt/neomutt/blob/master/mutt/notify_type.h#L31-L45), it should be clear.
Find where `my_hdr` and `unmy_hdr` are processed and send a global Event (to `NeoMutt->notify`).

Setting up the observing side should be pretty simple, e.g. 31d786e73

The final part will be to use the observed Event to update (add/remove) user headers from the current Envelope.

**Comment by matthewhughes934** [comment]
I've got a working (though still WIP) approach for redrawing the compose screen when 'my_hdr' is called under [devel/matthugh/my-hdr-notifications](https://github.com/neomutt/neomutt/tree/devel/matthugh/my-hdr-notifications). I need to give some thought into how best to handle sharing some logic currently in `parse_my_hdr` around finding a node in a header list and creating/updating as needed.

I'm yet to look into handling calls to 'unmy_hdr' here, but I think I will need to do some similar sharing for `parse_unmy_hdr` as I will do for `parse_my_hdr`, so hopefully once I have that done this will happen simply (famous last words)

**Comment by flatcap** [comment]
That's looking good!

**Comment by matthewhughes934** [comment]
@flatcap I'm wondering how best to implement the `unmy_hdr` notification+observer. I'm wanting to send a single notification per call, but this doesn't work too well with my current implementation of the `EventHeader` (see 504fdc073a25c67485b8b41fe793b8b9d28ac5e4) since this only takes a single header. My thought is to change this to:

```c
struct EventHeader
{
  char **headers;
  size_t num_headers;
};
```

Which would make implementing this observer simpler, but add some complication to the simpler `my_hdr` observers since there will have to be some extra allocating/freeing there. I'd be interested to hear your thoughts on it.

**Comment by flatcap** [comment]
I think keeping the notifications simple, is best.

Some situations, like Emails arriving, will need to send a large set of changes in one Event,
but the `my_hdr` notifications are going to be low volume.

So, one notification per header.

~Except for the `unmh_hdr *`, where I think we should send `NT_HEADER_REMOVE` + `EventHeader { NULL }`~

Edit: @matthewhughes934 pointed out that doesn't distinguish between the `my_hdr`s and those added by the user in the editor.

**Comment by matthewhughes934** [comment]
Thanks @flatcap. I've got something working on devel/matthugh/unmy-hdr-notifications (branches off devel/matthugh/my-hdr-notifications, I'll wait until #2649 is in a happy state before adding another PR)

**Comment by flatcap** [comment]
# Eliminating Config Globals

NeoMutt still has a lot of global variables.

Most of the Config Variables are backed by globals.
This is why when you switch Accounts you need to use `account-` and `folder-hook`s to set variables.

## Config System

The Config System support inheritance which will allow us to create Account- and Mailbox-specific config.
To use this facility means eliminating all of the `C_*` global variables.

This leads to my [favourite diagram](https://github.com/neomutt/gfx/blob/master/arch/README.md#config-and-inheritance) (and possibly the most complicated).
Understanding this will lead to Enlightenment :-)

## Config Subset

The Config System can be accessed using a handle: `struct ConfigSubset *sub`
Initially, this will be the **global scope**, `NeoMutt->sub`.
Over time, this will be changed to `Account->sub` when Account-specific config is introduced, then eventually `Mailbox->sub`.

For dialogs, the `ConfigSubset` should be passed in as a parameter.

## Conversion

To upgrade the code, replace each `C_*` global variable with a Config System lookup.
There's a helper function for each config type.
These are strictly type-checked and will `assert()` on failure.

Here's an example of each function.

```c
const struct Address  *c_envelope_from_address = cs_subset_address(sub, "envelope_from_address"); // C_EnvelopeFromAddress
const bool             c_fast_reply            = cs_subset_bool   (sub, "fast_reply");            // C_FastReply
const long             c_imap_fetch_chunk_size = cs_subset_long   (sub, "imap_fetch_chunk_size"); // C_ImapFetchChunkSize
const short            c_connect_timeout       = cs_subset_number (sub, "connect_timeout");       // C_ConnectTimeout
const char const      *c_debug_file            = cs_subset_path   (sub, "debug_file");            // C_DebugFile
const enum QuadOption  c_fcc_attach            = cs_subset_quad   (sub, "fcc_attach");            // C_FccAttach
const struct Regex    *c_gecos_mask            = cs_subset_regex  (sub, "gecos_mask");            // C_GecosMask
const struct Slist    *c_hidden_tags           = cs_subset_slist  (sub, "hidden_tags");           // C_HiddenTags
const short            c_pgp_sort_keys         = cs_subset_sort   (sub, "pgp_sort_keys");         // C_PgpSortKeys
const char const      *c_pattern_format        = cs_subset_string (sub, "pattern_format");        // C_PatternFormat
```

Each variable is named to match the Config Variable and is **const** to discourage the coder from changing it (which would have no effect on the actual config).

## Elimination

Once all the `C_*` variable uses have been eliminated from a library, the Config
Definitions can be updated.

- Delete the `C_*` config definitions
- Delete the `C_*` config prototypes
- Replace `C_*` with `NULL` in the Config Definition
- Flag the Config with `DT_NO_VARIABLE` in the call to `cs_register_variables()`

See 47a7ab79a for an example.

**Comment by flatcap** [comment]
For the `C_Sort` and `C_SortAux`, we can use the helpers to get the values,
then use `cs_subset_str_string_set()` to restore the values afterwards.

```c
int cs_subset_str_string_set(sub, "sort", old_sort, NULL);
```

---

That Compose code is about to open a new Index view so that the user can attach an Email.
(the call stack will look like: Index -\> Compose -> Index)

The **new** Index should be separate and distinct from the original.
In the **original** Index there's a 'current' Mailbox (the Context) and values for `$sort` and `$sort_aux`.
When we open the **new** Index, we want to see those same settings.
However...
If we change anything in the **new** Index -- change sorting or open a different Mailbox -- these changes must not alter the state of the **original** Index.

**Comment by flatcap** [comment]
Following on from IRC:
We already have the pointer we need
`snd_make_entry()` has a Menu, which gives us `Menu->redraw_data`, which is a `ComposeRedrawData`, which has `ComposeRedrawData->sub`

**Comment by flatcap** [comment]
Please can you replace the `NeoMutt->sub` with the `sub` param passed-in.
There are 15 of them. The 13 in `mutt_compose_menu()` are trivial.
`cum_attachs_size()` has a `Menu` and `check_attachments()` will need it passed in.

**Comment by flatcap** [comment]
To tick that last box is a bit tricky. I'm still thinking through the solution.

Briefly: I want to move the data out of functions and attach it to the window.

Below are diagrams showing the "Index-Compose-Index" situation.
Starting in the Index, Compose a new Email, then go to attach an Email, causing a second Index to open.

On the left is the current situation and on the right, the future plan.

**Key**:
- **Blue** - Windows
- **Yellow** - Functions
- **Purple** - Config
- **Grey** - Config Global Variable

<table>
  <tr>
    <td valign="top"><img width="450" src="https://flatcap.org/mutt/compose/c1.svg"></td>
    <td valign="top"><img width="450" src="https://flatcap.org/mutt/compose/c2.svg"></td>
  </tr>
  <tr>
    <td valign="top"><a href="https://flatcap.org/mutt/compose/c1.svg">svg source</a></td>
    <td valign="top"><a href="https://flatcap.org/mutt/compose/c2.svg">svg source</a></td>
  </tr>
</table>

### Left (Current)

On the left we see the Index functions.  `index()` has some data, which it passes to other functions.
If an Event were to happen outside of `index()`, then it wouldn't be able to see the data.

### Right (Future)

On the right, we've attached the data to the Windows.
The new `index()` function (not shown), would be passed a pointer to the Window (or its data).
External Events happening to the Windows would have access to the Window data.

This diagram shows Account-specific config.
It also shows Compose- and Index-specific config, which would be used to handle backgrounding Compose sessions.

---

**Note**: This is a work-in-progress; it **will** change :-)
